tokio/src/sync/broadcast.rs

```diff
@@ -1084,9 +1107,40 @@ impl<T> Receiver<T> {
                         return Err(TryRecvError::Closed);
                     }
 
+                    // We will might want to upgrade to a write lock.
+                    let mut tail_read = Some(tail);
+                    let mut tail_write = None;
+
                     // Store the waker
                     if let Some((waiter, waker)) = waiter {
-                        // Safety: called while locked.
+                        let queued = waiter.with(|ptr| {
+                            // Safety: waiter.queued is atomic.
+                            unsafe { (*ptr).queued.load(Relaxed) }
+                        });
+
+                        // Release the slot lock before reacquiring tail locks
+                        // to avoid a deadlock.
+                        drop(slot);
+
+                        // If waiter is already queued, then a write lock on tail is required
+                        // since other threads may try to mutate the waiter concurrently.
+                        // If the waiter is not queued, we are the only owner now and
+                        // read lock suffices.
+                        let tail_ref: &Tail = if queued {
+                            // TODO: this is sketchy, need do make sure that
+                            // it is safe to drop all the locks here...
+                            tail_read = None;
```

Comment:
I implemented the solution that we discussed above, and it gives a nice 15-25% performance boost on the benchmark on top of what was in the initial PR. I am worried about this place though: to upgrade read lock to a write lock we need to release all the locks. I will need to think more if it is safe.

Reply:
There was indeed a bug here, fixed.